### ISSUE TITLE
`--num-shard` causes memory leak

### ISSUE THREAD

**Issue by mwbyeon** [huggingface/text-generation-inference#655]
### System Info

* Model
```
{
    "model_id": "bigscience/bloom-7b1",
    "model_sha": "e83e90ba86f87f74aa2731cdab25ccf33976bd66",
    "model_dtype": "torch.float16",
    "model_device_type": "cuda",
    "model_pipeline_tag": "text-generation",
    "max_concurrent_requests": 128,
    "max_best_of": 2,
    "max_stop_sequences": 4,
    "max_input_length": 2048,
    "max_total_tokens": 4096,
    "waiting_served_ratio": 1.2,
    "max_batch_total_tokens": 16000,
    "max_waiting_tokens": 20,
    "validation_workers": 2,
    "version": "0.9.3",
    "sha": "5e6ddfd6a4fecc394255d7109f87c420c98b4e15",
    "docker_label": "sha-5e6ddfd"
}
```

* Environment: A100 40GBx2 Machine

### Information

- [X] Docker
- [ ] The CLI directly

### Tasks

- [X] An officially supported command
- [ ] My own modifications

### Reproduction

1. Run TGI with Docker
```
$ docker run -it --gpus all --shm-size 2g \
  -p 8081:80 \
  -v /home/dylan/tgi-data:/data \
  ghcr.io/huggingface/text-generation-inference:0.9.3 \
    --max-input-length=2048 \
    --num-shard=2 \
    --max-total-tokens=4096 \
    --disable-custom-kernels \
    --model-id=bigscience/bloom-7b1
```

2. Request using CURL
```
$ curl 127.0.0.1:8081/generate_stream   \
  -X POST   \
  -d '{"inputs":"What is Deep Learning?","parameters":{"max_new_tokens":1000, "num_beams": 1, "penalty_alpha": 0.6, "top_k": 8, "no_repeat_ngram_size": 6}}'   \
  -H 'Content-Type: application/json'
```

### Expected behavior

Each call to `/generate_stream` increments memory.

1. Initial memory usage
```
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 525.105.17   Driver Version: 525.105.17   CUDA Version: 12.0     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  NVIDIA A100-SXM...  Off  | 00000000:00:04.0 Off |                    0 |
| N/A   30C    P0    61W / 400W |  11791MiB / 40960MiB |      0%      Default |
|                               |                      |             Disabled |
+-------------------------------+----------------------+----------------------+
|   1  NVIDIA A100-SXM...  Off  | 00000000:00:05.0 Off |                    0 |
| N/A   38C    P0    64W / 400W |  11791MiB / 40960MiB |      0%      Default |
|                               |                      |             Disabled |
+-------------------------------+----------------------+----------------------+
```

2. Memory usage after 1 run
```
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 525.105.17   Driver Version: 525.105.17   CUDA Version: 12.0     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  NVIDIA A100-SXM...  Off  | 00000000:00:04.0 Off |                    0 |
| N/A   31C    P0    61W / 400W |  15823MiB / 40960MiB |      0%      Default |
|                               |                      |             Disabled |
+-------------------------------+----------------------+----------------------+
|   1  NVIDIA A100-SXM...  Off  | 00000000:00:05.0 Off |                    0 |
| N/A   33C    P0    76W / 400W |  15887MiB / 40960MiB |      0%      Default |
|                               |                      |             Disabled |
+-------------------------------+----------------------+----------------------+
```

3. Memory usage after 2 runs
```
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 525.105.17   Driver Version: 525.105.17   CUDA Version: 12.0     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  NVIDIA A100-SXM...  Off  | 00000000:00:04.0 Off |                    0 |
| N/A   32C    P0    62W / 400W |  19151MiB / 40960MiB |      0%      Default |
|                               |                      |             Disabled |
+-------------------------------+----------------------+----------------------+
|   1  NVIDIA A100-SXM...  Off  | 00000000:00:05.0 Off |                    0 |
| N/A   34C    P0    77W / 400W |  19215MiB / 40960MiB |      0%      Default |
|                               |                      |             Disabled |
+-------------------------------+----------------------+----------------------+
```

**Comment by Narsil** [comment]
Does it end up crashing ?

Since 0.9.2 (IIRC) we're using paged attention, meaning TGI uses it's own allocator which means what you're seeing here is not necessarily representative of actual RAM usage.

Does it end up blowing VRAM and oom after ~50 calls, or does it smooth out ?

**Comment by mwbyeon** [comment]
@Narsil 

* If `num_shards=1`, memory is not incremented.
* Memory continues to grow with each request, and then the process terminates with an OOM.
```
  File "/opt/conda/lib/python3.9/site-packages/text_generation_server/models/custom_modeling/bloom_modeling.py", line 398, in forward
    context_layer, present, attention_probs = self.compute_attention(
  File "/opt/conda/lib/python3.9/site-packages/text_generation_server/models/custom_modeling/bloom_modeling.py", line 306, in compute_attention
    value_layer = torch.cat((past_value, value_layer), dim=1)
torch.cuda.OutOfMemoryError: Allocation on device 0 would exceed allowed memory. (out of memory)
Currently allocated     : 9.64 GiB
Requested               : 1.49 MiB
Device limit            : 39.39 GiB
Free (according to CUDA): 3.06 MiB
PyTorch limit (set by user-supplied memory fraction)
                        : 39.39 GiB
```
* In v0.8.2, there is no memory increase.

**Comment by OlivierDehaene** [comment]
Yes it's an issue we had when using the `cudaMallocAsync` torch allocator at low throughput. This should be fixed on `0.9.3` and `latest` uses the normal torch allocator and does not have the issue at all.
Can you try one of these and report back if you still have this problem?

**Comment by mwbyeon** [comment]
@Narsil @OlivierDehaene 
The same issue occurs in `v0.9.3`. 
However, I checked with the `latest` version and it seems to be resolved with no increase in memory usage.

Thanks 👍

**Comment by shayan1897** [comment]
> Yes it's an issue we had when using the `cudaMallocAsync` torch allocator at low throughput. This should be fixed on `0.9.3` and `latest` uses the normal torch allocator and does not have the issue at all.
> Can you try one of these and report back if you still have this problem?

It’s still have the same issue using the latest as well.
